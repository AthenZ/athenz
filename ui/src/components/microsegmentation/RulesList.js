/*
 * Copyright The Athenz Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import React from 'react';
import styled from '@emotion/styled';
import Button from '../denali/Button';
import { MODAL_TIME_OUT } from '../constants/constants';
import RequestUtils from '../utils/RequestUtils';
import RuleTable from './RuleTable';
import AddSegmentation from './AddSegmentation';

const MembersSectionDiv = styled.div`
    margin: 20px;
`;

const AddContainerDiv = styled.div`
    padding-bottom: 20px;
    display: flex;
    justify-content: space-between;
    align-items: center;
    flex-flow: row nowrap;
    float: right;
`;

export default class RulesList extends React.Component {
    constructor(props) {
        super(props);
        this.api = props.api;
        this.state = {
            segmentationData: props.data,
            errorMessage: null,
            showAddSegmentation: false,
        };
        this.toggleAddSegmentation = this.toggleAddSegmentation.bind(this);
        this.closeModal = this.closeModal.bind(this);
        this.reloadData = this.reloadData.bind(this);
    }

    toggleAddSegmentation() {
        this.setState({
            showAddSegmentation: !this.state.showAddSegmentation,
        });
    }

    reloadData() {
        this.api
            .getInboundOutbound(this.props.domain)
            .then((data) => {
                this.setState({
                    segmentationData: data,
                    showAddSegmentation: false,
                    showAddStaticInstances: false,
                });
                setTimeout(
                    () =>
                        this.setState({
                            showSuccess: false,
                            successMessage: '',
                        }),
                    MODAL_TIME_OUT
                );
            })
            .catch((err) => {
                this.setState({
                    errorMessage: RequestUtils.xhrErrorCheckHelper(err),
                });
            });
    }

    closeModal() {
        this.setState({ showSuccess: null });
    }

    render() {
        const { domain } = this.props;
        let addSegmentationButton = '';
        let addSegmentation = this.state.showAddSegmentation ? (
            <AddSegmentation
                api={this.api}
                domain={this.props.domain}
                onSubmit={this.reloadData}
                onCancel={this.toggleAddSegmentation}
                _csrf={this.props._csrf}
                showAddSegment={this.state.showAddSegmentation}
                justificationRequired={this.props.isDomainAuditEnabled}
            />
        ) : (
            ''
        );

        addSegmentationButton = (
            <AddContainerDiv>
                <div>
                    <Button secondary onClick={this.toggleAddSegmentation}>
                        Add ACL Policy
                    </Button>
                    {addSegmentation}
                </div>
            </AddContainerDiv>
        );

        let showInbound = this.state.segmentationData.inbound.length > 0;
        let showOutbound = this.state.segmentationData.outbound.length > 0;

        return (
            <MembersSectionDiv data-testid='segmentation-data-list'>
                {addSegmentationButton}

                {showInbound ? (
                    <RuleTable
                        category={'inbound'}
                        domain={domain}
                        api={this.api}
                        _csrf={this.props._csrf}
                        onSubmit={this.reloadData}
                        data={this.state.segmentationData.inbound}
                        caption='Inbound'
                        justificationRequired={this.props.isDomainAuditEnabled}
                    />
                ) : null}
                <br />
                {showOutbound ? (
                    <RuleTable
                        category={'outbound'}
                        domain={domain}
                        api={this.api}
                        _csrf={this.props._csrf}
                        onSubmit={this.reloadData}
                        data={this.state.segmentationData.outbound}
                        caption='Outbound'
                        justificationRequired={this.props.isDomainAuditEnabled}
                    />
                ) : null}
                {!showInbound && !showOutbound ? (
                    <div>
                        Use the Add ACL Policy button to add inbound and
                        outbound rules.
                    </div>
                ) : null}
            </MembersSectionDiv>
        );
    }
}
