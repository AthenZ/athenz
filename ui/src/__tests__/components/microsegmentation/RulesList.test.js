/*
 * Copyright The Athenz Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import React from 'react';
import { render } from '@testing-library/react';
import API from '../../../api';
import RulesList from "../../../components/microsegmentation/RulesList";

describe('RulesList', () => {
    it('should render', () => {
        let domain= 'domain';
        let segmentationData = {
            "inbound": [
                {
                    "source_service": "serviceA",
                    "source_port": "1111",
                    "destination_port": "2222",
                    "destination_services": [
                        "serviceB",
                        "serviceC",
                        "serviceD"
                    ],
                    "layer" : "tcp"
                }
            ],
            "outbound": [
                {
                    "source_service": "serviceE",
                    "source_port": "3333",
                    "destination_port": "4444",
                    "destination_services": [
                        "serviceF",
                        "serviceG",
                        "serviceH"
                    ],
                    "layer" : "udp"
                }
            ]
        };
        const { getByTestId } = render(
            <RulesList
                api={API()}
                domain={domain}
                _csrf={"_csrf"}
                isDomainAuditEnabled={true}
                data={segmentationData}
            />
        );
        const rulesList = getByTestId('segmentation-data-list');

        expect(rulesList).toMatchSnapshot();
    });
});
