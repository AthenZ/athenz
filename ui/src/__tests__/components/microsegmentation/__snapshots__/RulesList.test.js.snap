// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RulesList should render 1`] = `
.emotion-0 {
  margin: 20px;
}

.emotion-2 {
  padding-bottom: 20px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-flex-flow: row nowrap;
  -webkit-flex-flow: row nowrap;
  -ms-flex-flow: row nowrap;
  flex-flow: row nowrap;
  float: right;
}

.emotion-4 {
  border: none;
  border-radius: 2px;
  box-shadow: 0 0 0 1px transparent inset,0 0 0 0 #d5d5d5 inset;
  box-sizing: border-box;
  cursor: pointer;
  display: inline-block;
  font-family: Helvetica,Arial,sans-serif;
  font-weight: 300;
  font-size: 14px;
  line-height: 1;
  margin: 5px;
  outline: 0;
  padding: 10px 24px;
  text-align: center;
  text-shadow: none;
  text-transform: none;
  -webkit-transition: all 0.2s ease-in;
  transition: all 0.2s ease-in;
  vertical-align: baseline;
  white-space: nowrap;
  background: transparent;
  box-shadow: 0 0 0 1px #3697f2 inset;
  color: #3697f2;
}

.emotion-4:first-of-type {
  margin-left: 0;
}

.emotion-4:disabled {
  background: rgba(48,48,48,0.1);
  color: rgba(48,48,48,0.2);
  cursor: not-allowed;
}

.emotion-4:disabled {
  box-shadow: 0 0 0 1px rgba(48,48,48,0.1) inset;
}

.emotion-4:hover:not(:disabled) {
  background: #d7e2fd;
}

.emotion-5 {
  width: 100%;
  border-spacing: 0;
  display: table;
  border-collapse: separate;
  border-color: grey;
  box-sizing: border-box;
  margin-top: 5px;
  box-shadow: 0 1px 4px #d9d9d9;
  border: 1px solid #fff;
  -webkit-border-image: none;
  border-image: none;
  -webkit-border-image: initial;
  border-image: initial;
  height: 50px;
}

.emotion-7 {
  height: 25px;
  margin-left: 10px;
  margin-top: 10px;
  text-align: left;
  padding: 5px 0 5px 15px;
  vertical-align: middle;
  word-break: break-all;
  display: block;
}

.emotion-9 {
  margin-right: 10px;
  verticalalign: bottom;
}

.emotion-11 {
  fill: #188fff;
  cursor: pointer;
  vertical-align: text-bottom;
}

.emotion-12 {
  text-align: left;
  border-bottom: 2px solid #d5d5d5;
  color: #9a9a9a;
  font-weight: 600;
  font-size: 0.8rem;
  padding-bottom: 5px;
  vertical-align: top;
  text-transform: uppercase;
  padding: 5px 0 5px 35px;
  word-break: break-all;
}

.emotion-16 {
  text-align: left;
  border-bottom: 2px solid #d5d5d5;
  color: #9a9a9a;
  font-weight: 600;
  font-size: 0.8rem;
  padding-bottom: 5px;
  vertical-align: top;
  text-transform: uppercase;
  padding: 5px 0 5px 15px;
  word-break: break-all;
}

.emotion-24 {
  text-align: center;
  border-bottom: 2px solid #d5d5d5;
  color: #9a9a9a;
  font-weight: 600;
  font-size: 0.8rem;
  padding-bottom: 5px;
  vertical-align: top;
  text-transform: uppercase;
  padding: 5px 0 5px 15px;
  word-break: break-all;
}

.emotion-32 {
  background-color: #3570f40D;
  text-align: left;
  padding: 5px 0 5px 15px;
  vertical-align: middle;
  word-break: break-all;
}

.emotion-38 {
  background-color: #3570f40D;
  text-align: left;
  padding: 5px 0 5px 15px;
  vertical-align: middle;
  word-break: break-all;
  -webkit-text-decoration: dashed underline;
  text-decoration: dashed underline;
}

.emotion-45 {
  background-color: #3570f40D;
  text-align: center;
  padding: 5px 0 5px 15px;
  vertical-align: middle;
  word-break: break-all;
  -webkit-text-decoration: dashed underline;
  text-decoration: dashed underline;
}

.emotion-48 {
  background-color: #3570f40D;
  text-align: center;
  padding: 5px 0 5px 15px;
  vertical-align: middle;
  word-break: break-all;
}

<div
  class="emotion-0 emotion-1"
  data-testid="segmentation-data-list"
>
  <div
    class="emotion-2 emotion-3"
  >
    <div>
      <button
        class="emotion-4 denali-button"
      >
        Add ACL Policy
      </button>
      
    </div>
  </div>
  <table
    class="emotion-5 emotion-6"
    data-testid="segmentation-rule-table"
  >
    <thead>
      <caption
        class="emotion-7 emotion-8"
      >
        <span
          class="emotion-9 emotion-10"
        >
          <svg
            class="emotion-11"
            data-testid="icon"
            height="1.25em"
            viewBox="0 0 1024 1024"
            width="1.25em"
          >
            <title>
              arrowhead-up-circle-solid
            </title>
            <path
              d="M512 42.667c-259.206 0-469.333 210.128-469.333 469.333s210.128 469.333 469.333 469.333c259.206 0 469.333-210.128 469.333-469.333v0c0.003-0.635 0.005-1.387 0.005-2.138 0-258.027-209.173-467.2-467.2-467.2-0.752 0-1.504 0.002-2.255 0.005l0.117-0zM718.933 601.6c-9.129 9.862-21.88 16.257-36.127 17.060l-0.14 0.006h-341.333c-14.386-0.81-27.138-7.205-36.235-17.032l-0.032-0.035c-4.299-6.754-6.851-14.984-6.851-23.81 0-13.46 5.936-25.533 15.332-33.745l0.053-0.045 168.533-168.533c7.434-8.083 18.061-13.13 29.867-13.13s22.432 5.047 29.84 13.101l0.026 0.029 168.533 168.533c9.449 8.257 15.384 20.33 15.384 33.79 0 8.826-2.552 17.056-6.959 23.992l0.108-0.182z"
            />
          </svg>
        </span>
        Inbound (1)
      </caption>
      <tr>
        <th
          class="emotion-12 emotion-13"
        >
          Identifier
        </th>
        <th
          class="emotion-12 emotion-13"
        >
          Destination Service
        </th>
        <th
          class="emotion-16 emotion-17"
        >
          Destination Port
        </th>
        <th
          class="emotion-16 emotion-17"
        >
          Source Service
        </th>
        <th
          class="emotion-16 emotion-17"
        >
          Source Port
        </th>
        <th
          class="emotion-16 emotion-17"
        >
          Layer
        </th>
        <th
          class="emotion-24 emotion-17"
        >
          Enforcement State
        </th>
        <th
          class="emotion-24 emotion-17"
        >
          Edit
        </th>
        <th
          class="emotion-24 emotion-17"
        >
          Delete
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        class="emotion-30 emotion-31"
        data-testid="segmentation-row"
      >
        <td
          class="emotion-32 emotion-33"
          color="#3570f40D"
        />
        <td
          class="emotion-32 emotion-33"
          color="#3570f40D"
        />
        <td
          class="emotion-32 emotion-33"
          color="#3570f40D"
        >
          2222
        </td>
        <td
          class="emotion-38 emotion-39"
          color="#3570f40D"
        >
          <span>
            <svg
              class="emotion-11"
              data-testid="icon"
              height="1.25em"
              viewBox="0 0 1024 1024"
              width="1.25em"
            >
              <title>
                service-setting
              </title>
              <path
                d="M612.267 401.067l-38.4 61.867 29.867 29.867 61.867-38.4 21.333 8.533 17.067 70.4h42.667l17.067-70.4 21.333-8.533 61.867 38.4 29.867-29.867-38.4-61.867c3.496-6.133 6.416-13.229 8.393-20.708l0.14-0.625 70.4-17.067v-42.667l-70.4-17.067c-2.117-8.104-5.037-15.2-8.765-21.778l0.232 0.445 38.4-61.867-29.867-29.867-61.867 38.4-21.333-8.533-17.067-70.4h-42.667l-17.067 70.4-21.333 8.533-61.867-38.4-29.867 29.867 38.4 61.867c-3.496 6.133-6.416 13.229-8.393 20.708l-0.14 0.625-70.4 17.067v42.667l70.4 17.067c2.117 8.104 5.037 15.2 8.765 21.778l-0.232-0.445zM725.333 298.667c23.564 0 42.667 19.103 42.667 42.667s-19.103 42.667-42.667 42.667c-23.564 0-42.667-19.103-42.667-42.667v0c0-23.564 19.103-42.667 42.667-42.667v0z"
              />
              <path
                d="M900.267 612.267c-18.864-15.311-43.169-24.582-69.64-24.582-5.536 0-10.978 0.406-16.297 1.189l0.603-0.073-211.2 38.4c0.002-0.241 0.003-0.525 0.003-0.81 0-27.37-11.454-52.065-29.83-69.553l-0.040-0.038-151.467-151.467h-315.733c-23.564 0-42.667 19.103-42.667 42.667s19.103 42.667 42.667 42.667v0h281.6l125.867 125.867c3.276 3.276 5.302 7.801 5.302 12.8 0 9.997-8.105 18.102-18.102 18.102-4.999 0-9.524-2.026-12.8-5.302l-74.667-74.667c-7.042-5.327-15.947-8.533-25.6-8.533-23.564 0-42.667 19.103-42.667 42.667 0 9.653 3.206 18.558 8.611 25.707l-0.078-0.107 74.667 74.667c17.528 18.44 42.239 29.911 69.629 29.911 1.021 0 2.039-0.016 3.052-0.048l-0.148 0.004c13.761-0.062 26.955-2.401 39.254-6.657l-0.854 0.257 290.133-51.2c1.148-0.219 2.469-0.345 3.82-0.345 5.025 0 9.645 1.738 13.29 4.645l-0.043-0.033c3.941 3.531 6.41 8.637 6.41 14.32 0 0.216-0.004 0.431-0.011 0.645l0.001-0.031c0.003 0.128 0.004 0.279 0.004 0.43 0 10.273-7.262 18.851-16.932 20.879l-0.139 0.024-396.8 72.533-232.533-189.867h-100.267c-23.564 0-42.667 19.103-42.667 42.667s19.103 42.667 42.667 42.667v0h70.4l236.8 194.133 437.333-78.933c49.475-9.916 86.226-53.011 86.226-104.69 0-32.293-14.35-61.234-37.022-80.794l-0.138-0.116z"
              />
            </svg>
          </span>
        </td>
        <td
          class="emotion-32 emotion-33"
          color="#3570f40D"
        >
          1111
        </td>
        <td
          class="emotion-32 emotion-33"
          color="#3570f40D"
        >
          tcp
        </td>
        <td
          class="emotion-45 emotion-39"
          color="#3570f40D"
        >
          <span>
            <svg
              class="emotion-11"
              data-testid="icon"
              height="1.25em"
              viewBox="0 0 1024 1024"
              width="1.25em"
            >
              <title>
                network-resource
              </title>
              <path
                d="M170.667 213.333h388.267l29.867 29.867-55.467 55.467h170.667v-170.667l-55.467 55.467-55.467-55.467h-422.4c-23.564 0-42.667 19.103-42.667 42.667s19.103 42.667 42.667 42.667v0z"
              />
              <path
                d="M588.8 780.8l-29.867 29.867h-388.267c-23.564 0-42.667 19.103-42.667 42.667s19.103 42.667 42.667 42.667v0h422.4l55.467-55.467 55.467 55.467v-170.667h-170.667z"
              />
              <path
                d="M170.667 554.667h256v85.333l128-128-128-128v85.333h-256c-23.564 0-42.667 19.103-42.667 42.667s19.103 42.667 42.667 42.667v0z"
              />
              <path
                d="M768 341.333c-94.257 0-170.667 76.41-170.667 170.667s76.41 170.667 170.667 170.667c94.257 0 170.667-76.41 170.667-170.667v0c0-94.257-76.41-170.667-170.667-170.667v0zM768 597.333c-47.128 0-85.333-38.205-85.333-85.333s38.205-85.333 85.333-85.333c47.128 0 85.333 38.205 85.333 85.333v0c0 47.128-38.205 85.333-85.333 85.333v0z"
              />
            </svg>
          </span>
        </td>
        <td
          class="emotion-48 emotion-33"
          color="#3570f40D"
        >
          <span>
            <svg
              class="emotion-11"
              data-testid="icon"
              height="1.25em"
              viewBox="0 0 1024 1024"
              width="1.25em"
            >
              <title>
                pencil
              </title>
              <path
                d="M925.867 401.067l-300.8-302.933c-7.434-8.083-18.061-13.13-29.867-13.13s-22.432 5.047-29.84 13.101l-0.026 0.029-467.2 467.2c-7.763 7.612-12.628 18.155-12.8 29.835l-0 0.032v300.8c0 23.564 19.103 42.667 42.667 42.667v0h300.8c11.712-0.172 22.255-5.037 29.86-12.793l0.007-0.007 467.2-465.067c8.083-7.434 13.13-18.061 13.13-29.867s-5.047-22.432-13.101-29.84l-0.029-0.026zM411.733 853.333h-241.067v-241.067l266.667-264.533 238.933 241.067zM738.133 529.067l-241.067-241.067 98.133-100.267 241.067 243.2z"
              />
            </svg>
          </span>
        </td>
        
        <td
          class="emotion-48 emotion-33"
          color="#3570f40D"
        >
          <span>
            <svg
              class="emotion-11"
              data-testid="icon"
              height="1.25em"
              viewBox="0 0 1024 1024"
              width="1.25em"
            >
              <title>
                trash
              </title>
              <path
                d="M917.333 187.733c0.007 0.319 0.012 0.694 0.012 1.071 0 12.755-4.867 24.374-12.846 33.1l0.034-0.038c-6.696 7.89-16.62 12.864-27.706 12.864-0.76 0-1.515-0.023-2.264-0.069l0.103 0.005h-256c-23.564 0-42.667-19.103-42.667-42.667v0-42.667h-128v42.667c0 23.564-19.103 42.667-42.667 42.667v0h-253.867c-0.136 0.001-0.298 0.002-0.459 0.002-22.485 0-41.102-16.565-44.311-38.157l-0.030-0.245c-0.007-0.319-0.012-0.694-0.012-1.071 0-12.755 4.867-24.374 12.846-33.1l-0.034 0.038c6.696-7.89 16.62-12.864 27.706-12.864 0.76 0 1.515 0.023 2.264 0.069l-0.103-0.005h213.333v-42.667c0-23.564 19.103-42.667 42.667-42.667v0h213.333c23.564 0 42.667 19.103 42.667 42.667v0 42.667h211.2c0.136-0.001 0.298-0.002 0.459-0.002 22.485 0 41.102 16.565 44.311 38.157l0.030 0.245z"
              />
              <path
                d="M742.4 358.4l-51.2 516.267h-358.4l-51.2-516.267c-2.24-21.656-20.391-38.401-42.453-38.401-0.075 0-0.15 0-0.225 0.001l0.012-0c-0.064-0-0.139-0.001-0.214-0.001-23.564 0-42.667 19.103-42.667 42.667 0 1.503 0.078 2.988 0.229 4.45l-0.015-0.183 55.467 554.667c2.24 21.656 20.391 38.401 42.453 38.401 0.075 0 0.15-0 0.225-0.001l-0.012 0h435.2c0.064 0 0.139 0.001 0.214 0.001 22.062 0 40.213-16.744 42.437-38.218l0.015-0.183 55.467-554.667c0.136-1.28 0.214-2.764 0.214-4.267 0-23.564-19.103-42.667-42.667-42.667-0.075 0-0.15 0-0.226 0.001l0.012-0c-0.064-0-0.139-0.001-0.214-0.001-22.062 0-40.213 16.744-42.437 38.218l-0.015 0.183z"
              />
            </svg>
          </span>
        </td>
      </tr>
    </tbody>
  </table>
  <br />
  <table
    class="emotion-5 emotion-6"
    data-testid="segmentation-rule-table"
  >
    <thead>
      <caption
        class="emotion-7 emotion-8"
      >
        <span
          class="emotion-9 emotion-10"
        >
          <svg
            class="emotion-11"
            data-testid="icon"
            height="1.25em"
            viewBox="0 0 1024 1024"
            width="1.25em"
          >
            <title>
              arrowhead-up-circle-solid
            </title>
            <path
              d="M512 42.667c-259.206 0-469.333 210.128-469.333 469.333s210.128 469.333 469.333 469.333c259.206 0 469.333-210.128 469.333-469.333v0c0.003-0.635 0.005-1.387 0.005-2.138 0-258.027-209.173-467.2-467.2-467.2-0.752 0-1.504 0.002-2.255 0.005l0.117-0zM718.933 601.6c-9.129 9.862-21.88 16.257-36.127 17.060l-0.14 0.006h-341.333c-14.386-0.81-27.138-7.205-36.235-17.032l-0.032-0.035c-4.299-6.754-6.851-14.984-6.851-23.81 0-13.46 5.936-25.533 15.332-33.745l0.053-0.045 168.533-168.533c7.434-8.083 18.061-13.13 29.867-13.13s22.432 5.047 29.84 13.101l0.026 0.029 168.533 168.533c9.449 8.257 15.384 20.33 15.384 33.79 0 8.826-2.552 17.056-6.959 23.992l0.108-0.182z"
            />
          </svg>
        </span>
        Outbound (1)
      </caption>
      <tr>
        <th
          class="emotion-12 emotion-13"
        >
          Identifier
        </th>
        <th
          class="emotion-12 emotion-13"
        >
          Source Service
        </th>
        <th
          class="emotion-16 emotion-17"
        >
          Source Port
        </th>
        <th
          class="emotion-16 emotion-17"
        >
          Destination Service
        </th>
        <th
          class="emotion-16 emotion-17"
        >
          Destination Port
        </th>
        <th
          class="emotion-16 emotion-17"
        >
          Layer
        </th>
        <th
          class="emotion-24 emotion-17"
        >
          Enforcement State
        </th>
        <th
          class="emotion-24 emotion-17"
        >
          Edit
        </th>
        <th
          class="emotion-24 emotion-17"
        >
          Delete
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        class="emotion-30 emotion-31"
        data-testid="segmentation-row"
      >
        <td
          class="emotion-32 emotion-33"
          color="#3570f40D"
        />
        <td
          class="emotion-32 emotion-33"
          color="#3570f40D"
        >
          serviceE
        </td>
        <td
          class="emotion-32 emotion-33"
          color="#3570f40D"
        >
          3333
        </td>
        <td
          class="emotion-38 emotion-39"
          color="#3570f40D"
        >
          <span>
            <svg
              class="emotion-11"
              data-testid="icon"
              height="1.25em"
              viewBox="0 0 1024 1024"
              width="1.25em"
            >
              <title>
                service-setting
              </title>
              <path
                d="M612.267 401.067l-38.4 61.867 29.867 29.867 61.867-38.4 21.333 8.533 17.067 70.4h42.667l17.067-70.4 21.333-8.533 61.867 38.4 29.867-29.867-38.4-61.867c3.496-6.133 6.416-13.229 8.393-20.708l0.14-0.625 70.4-17.067v-42.667l-70.4-17.067c-2.117-8.104-5.037-15.2-8.765-21.778l0.232 0.445 38.4-61.867-29.867-29.867-61.867 38.4-21.333-8.533-17.067-70.4h-42.667l-17.067 70.4-21.333 8.533-61.867-38.4-29.867 29.867 38.4 61.867c-3.496 6.133-6.416 13.229-8.393 20.708l-0.14 0.625-70.4 17.067v42.667l70.4 17.067c2.117 8.104 5.037 15.2 8.765 21.778l-0.232-0.445zM725.333 298.667c23.564 0 42.667 19.103 42.667 42.667s-19.103 42.667-42.667 42.667c-23.564 0-42.667-19.103-42.667-42.667v0c0-23.564 19.103-42.667 42.667-42.667v0z"
              />
              <path
                d="M900.267 612.267c-18.864-15.311-43.169-24.582-69.64-24.582-5.536 0-10.978 0.406-16.297 1.189l0.603-0.073-211.2 38.4c0.002-0.241 0.003-0.525 0.003-0.81 0-27.37-11.454-52.065-29.83-69.553l-0.040-0.038-151.467-151.467h-315.733c-23.564 0-42.667 19.103-42.667 42.667s19.103 42.667 42.667 42.667v0h281.6l125.867 125.867c3.276 3.276 5.302 7.801 5.302 12.8 0 9.997-8.105 18.102-18.102 18.102-4.999 0-9.524-2.026-12.8-5.302l-74.667-74.667c-7.042-5.327-15.947-8.533-25.6-8.533-23.564 0-42.667 19.103-42.667 42.667 0 9.653 3.206 18.558 8.611 25.707l-0.078-0.107 74.667 74.667c17.528 18.44 42.239 29.911 69.629 29.911 1.021 0 2.039-0.016 3.052-0.048l-0.148 0.004c13.761-0.062 26.955-2.401 39.254-6.657l-0.854 0.257 290.133-51.2c1.148-0.219 2.469-0.345 3.82-0.345 5.025 0 9.645 1.738 13.29 4.645l-0.043-0.033c3.941 3.531 6.41 8.637 6.41 14.32 0 0.216-0.004 0.431-0.011 0.645l0.001-0.031c0.003 0.128 0.004 0.279 0.004 0.43 0 10.273-7.262 18.851-16.932 20.879l-0.139 0.024-396.8 72.533-232.533-189.867h-100.267c-23.564 0-42.667 19.103-42.667 42.667s19.103 42.667 42.667 42.667v0h70.4l236.8 194.133 437.333-78.933c49.475-9.916 86.226-53.011 86.226-104.69 0-32.293-14.35-61.234-37.022-80.794l-0.138-0.116z"
              />
            </svg>
          </span>
        </td>
        <td
          class="emotion-32 emotion-33"
          color="#3570f40D"
        >
          4444
        </td>
        <td
          class="emotion-32 emotion-33"
          color="#3570f40D"
        >
          udp
        </td>
        <td
          class="emotion-45 emotion-39"
          color="#3570f40D"
        >
          <span>
            <svg
              class="emotion-11"
              data-testid="icon"
              height="1.25em"
              viewBox="0 0 1024 1024"
              width="1.25em"
            >
              <title>
                network-role
              </title>
              <path
                d="M896 512l-128-128v85.333h-347.733c-6.016-26.402-17.927-49.416-34.309-68.476l0.176 0.209 185.6-187.733h196.267v85.333l128-128-128-128v85.333h-230.4l-215.467 215.467-6.4 8.533c-17.75-6.504-38.245-10.39-59.61-10.665l-0.124-0.001c-94.257 0-170.667 76.41-170.667 170.667s76.41 170.667 170.667 170.667v0c21.488-0.277 41.983-4.163 61.025-11.077l-1.292 0.41 6.4 8.533 215.467 217.6h230.4v83.2l128-128-128-128v87.467h-196.267l-185.6-189.867c15.307-19.245 27.062-42.071 33.828-66.951l0.305-1.316h347.733v85.333zM256 597.333c-47.128 0-85.333-38.205-85.333-85.333s38.205-85.333 85.333-85.333c47.128 0 85.333 38.205 85.333 85.333v0c0 47.128-38.205 85.333-85.333 85.333v0z"
              />
            </svg>
          </span>
        </td>
        <td
          class="emotion-48 emotion-33"
          color="#3570f40D"
        >
          <span>
            <svg
              class="emotion-11"
              data-testid="icon"
              height="1.25em"
              viewBox="0 0 1024 1024"
              width="1.25em"
            >
              <title>
                pencil
              </title>
              <path
                d="M925.867 401.067l-300.8-302.933c-7.434-8.083-18.061-13.13-29.867-13.13s-22.432 5.047-29.84 13.101l-0.026 0.029-467.2 467.2c-7.763 7.612-12.628 18.155-12.8 29.835l-0 0.032v300.8c0 23.564 19.103 42.667 42.667 42.667v0h300.8c11.712-0.172 22.255-5.037 29.86-12.793l0.007-0.007 467.2-465.067c8.083-7.434 13.13-18.061 13.13-29.867s-5.047-22.432-13.101-29.84l-0.029-0.026zM411.733 853.333h-241.067v-241.067l266.667-264.533 238.933 241.067zM738.133 529.067l-241.067-241.067 98.133-100.267 241.067 243.2z"
              />
            </svg>
          </span>
        </td>
        
        <td
          class="emotion-48 emotion-33"
          color="#3570f40D"
        >
          <span>
            <svg
              class="emotion-11"
              data-testid="icon"
              height="1.25em"
              viewBox="0 0 1024 1024"
              width="1.25em"
            >
              <title>
                trash
              </title>
              <path
                d="M917.333 187.733c0.007 0.319 0.012 0.694 0.012 1.071 0 12.755-4.867 24.374-12.846 33.1l0.034-0.038c-6.696 7.89-16.62 12.864-27.706 12.864-0.76 0-1.515-0.023-2.264-0.069l0.103 0.005h-256c-23.564 0-42.667-19.103-42.667-42.667v0-42.667h-128v42.667c0 23.564-19.103 42.667-42.667 42.667v0h-253.867c-0.136 0.001-0.298 0.002-0.459 0.002-22.485 0-41.102-16.565-44.311-38.157l-0.030-0.245c-0.007-0.319-0.012-0.694-0.012-1.071 0-12.755 4.867-24.374 12.846-33.1l-0.034 0.038c6.696-7.89 16.62-12.864 27.706-12.864 0.76 0 1.515 0.023 2.264 0.069l-0.103-0.005h213.333v-42.667c0-23.564 19.103-42.667 42.667-42.667v0h213.333c23.564 0 42.667 19.103 42.667 42.667v0 42.667h211.2c0.136-0.001 0.298-0.002 0.459-0.002 22.485 0 41.102 16.565 44.311 38.157l0.030 0.245z"
              />
              <path
                d="M742.4 358.4l-51.2 516.267h-358.4l-51.2-516.267c-2.24-21.656-20.391-38.401-42.453-38.401-0.075 0-0.15 0-0.225 0.001l0.012-0c-0.064-0-0.139-0.001-0.214-0.001-23.564 0-42.667 19.103-42.667 42.667 0 1.503 0.078 2.988 0.229 4.45l-0.015-0.183 55.467 554.667c2.24 21.656 20.391 38.401 42.453 38.401 0.075 0 0.15-0 0.225-0.001l-0.012 0h435.2c0.064 0 0.139 0.001 0.214 0.001 22.062 0 40.213-16.744 42.437-38.218l0.015-0.183 55.467-554.667c0.136-1.28 0.214-2.764 0.214-4.267 0-23.564-19.103-42.667-42.667-42.667-0.075 0-0.15 0-0.226 0.001l0.012-0c-0.064-0-0.139-0.001-0.214-0.001-22.062 0-40.213 16.744-42.437 38.218l-0.015 0.183z"
              />
            </svg>
          </span>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
